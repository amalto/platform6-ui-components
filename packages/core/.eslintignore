dist/
loader/

src/components.d.ts