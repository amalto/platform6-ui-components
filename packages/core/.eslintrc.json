{
  "extends": ["../../.eslintrc.json", "plugin:@stencil/recommended", "plugin:jest/recommended", "plugin:jsx-a11y/recommended"],
  "parserOptions": {
    "project": ["./tsconfig.json", "./tsconfig.lint.json"],
    "ecmaFeatures": {
      "jsx": true // Allows for the parsing of JSX
    }
  },
  "plugins": ["jest"],

  "rules": {
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "assert": "either"
      }
    ],
    "react/jsx-props-no-spreading": "off",
    "react/no-unknown-property": [2, { "ignore": ["class"] }]
  },
  "settings": {
    "react": {
      "pragma": "h"
    }
  }
}
