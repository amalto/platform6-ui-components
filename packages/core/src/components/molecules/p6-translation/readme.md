# p6-translation-field

<!-- Auto Generated Below -->

## Properties

| Property            | Attribute  | Description                                                                               | Type                         | Default     |
| ------------------- | ---------- | ----------------------------------------------------------------------------------------- | ---------------------------- | ----------- |
| `disabled`          | `disabled` | the input is not available for interaction. The value will not be submitted with the form | `boolean`                    | `false`     |
| `name` _(required)_ | `name`     | The name                                                                                  | `string`                     | `undefined` |
| `readOnly`          | `readonly` | Marks as read only.                                                                       | `boolean`                    | `false`     |
| `value`             | --         | The value                                                                                 | `{ [key: string]: string; }` | `{}`        |

## Events

| Event              | Description                       | Type                                                |
| ------------------ | --------------------------------- | --------------------------------------------------- |
| `p6FormRegister`   | When the select ask to register   | `CustomEvent<P6Control<P6TranslationValue>>`        |
| `p6FormUnregister` | When the select ask to unregister | `CustomEvent<P6Control<P6TranslationValue>>`        |
| `p6Invalid`        | When the field is invalid         | `CustomEvent<InvalidEventDetail>`                   |
| `p6Valid`          | When the field is valid           | `CustomEvent<ValidEventDetail<P6TranslationValue>>` |

## Methods

### `checkValidity() => Promise<boolean>`

Returns whether a form will validate when it is submitted, without having to submit it.

#### Returns

Type: `Promise<boolean>`

### `validationMessage() => Promise<string>`

Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
It also triggers the standard error message, such as "this is a required field".

#### Returns

Type: `Promise<string>`

## Dependencies

### Depends on

- [p6-field](../p6-field)
- [p6-label](../../atoms/p6-label)
- [p6-action](../../atoms/p6-action)
- [p6-icon](../../atoms/p6-icon)
- [p6-translation-line](./components/p6-translation-line)

### Graph

```mermaid
graph TD;
  p6-translation --> p6-field
  p6-translation --> p6-label
  p6-translation --> p6-action
  p6-translation --> p6-icon
  p6-translation --> p6-translation-line
  p6-field --> p6-hint
  p6-translation-line --> p6-action
  p6-translation-line --> p6-icon
  p6-translation-line --> p6-language
  p6-translation-line --> p6-textarea
  p6-language --> p6-select-native
  style p6-translation fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
