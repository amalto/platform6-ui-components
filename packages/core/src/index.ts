export { Components, JSX } from './components';
export { P6ButtonType } from './components/atoms/p6-button/p6-button';
export { P6CalendarType } from './components/atoms/p6-calendar/p6-calendar';
export { P6CheckboxValue } from './components/atoms/p6-checkbox/p6-checkbox';
export { P6InputValue } from './components/atoms/p6-input/p6-input';
export { P6InputType } from './components/atoms/p6-input/types';
export { Target } from './components/atoms/p6-link/p6-link';
export { P6RadioValue } from './components/atoms/p6-radio/p6-radio';
export { P6SelectControl, P6SelectValue } from './components/atoms/p6-select-native/p6-select-native';
export { P6SwitchValue } from './components/atoms/p6-switch/p6-switch';
export { P6TextareaValue } from './components/atoms/p6-textarea/p6-textarea';
export { P6LanguageValue } from './components/molecules/p6-language/p6-language';
export { P6TranslationControl, P6TranslationValue } from './components/molecules/p6-translation/p6-translation';
export {
  AlignColumnDetail,
  CellValueChangedDetail,
  Column,
  ColumnDefinition,
  ColumnId,
  DataItem,
  EditingCellDetail,
  FilterRowsDetail,
  MoveColumnDetail,
  ResetDefinitionsDetail,
  ResizeColumnDetail,
  Row,
  RowId,
  ShowColumnDetail,
  ShowOptionsDetail,
  SortColumnDetail,
} from './components/organisms/p6-grid/core/entities';
export { ServiceStatus } from './components/templates/p6-service-panel/service-status.enum';
export { P6Control } from './shared/form/control';
export { InvalidEventDetail, isCustomEvent, isInvalidEvent, isValidEvent, ValidEventDetail } from './shared/form/event';
export { Alignment, Mode, modes, Position, positions, Size, sizes, SortOrder } from './shared/types';
export { EnumArrayEntry, enumArrayToObject, enumToArray, EnumType } from './utils/enum';
export { LanguageCode } from './utils/language';
