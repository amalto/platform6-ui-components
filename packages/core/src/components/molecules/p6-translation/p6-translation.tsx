import { Component, ComponentInterface, Element, Event, EventEmitter, h, Host, JSX, Method, Prop, State } from '@stencil/core';
import { P6Control } from '../../../shared/form/control';
import { InvalidEventDetail, isCustomEvent, ValidEventDetail } from '../../../shared/form/event';
import { Mode, Size } from '../../../shared/types';
import { getDefaultLanguage, getLanguageCodes, getLanguageName, LanguageCode } from '../../../utils/language';
import { getL10n, interpolate, L10n } from '../../../utils/translations';

export type Translation = { language: LanguageCode; value: string };
export type P6TranslationValue = { [key: string]: string };
export type P6TranslationControl = P6Control<P6TranslationValue>;

@Component({
  tag: 'p6-translation',
  styleUrl: 'p6-translation.scss',
  assetsDirs: ['locales'],
  shadow: true,
})
export class P6Translation implements ComponentInterface, P6TranslationControl {
  /**
   * The name
   */
  @Prop() name!: string;

  /**
   * The value
   */
  @Prop() value: P6TranslationValue = {};

  /**
   * Marks as read only.
   */
  @Prop({ attribute: 'readOnly' }) public readOnly = false;

  /**
   * the input is not available for interaction. The value will not be submitted with the form
   */
  @Prop() disabled = false;

  /**
   * When the field is valid
   */
  @Event() p6Valid!: EventEmitter<ValidEventDetail<P6TranslationValue>>;

  /**
   * When the field is invalid
   */
  @Event() p6Invalid!: EventEmitter<InvalidEventDetail>;

  /**
   * When the select ask to register
   */
  @Event() p6FormRegister!: EventEmitter<P6TranslationControl>;

  /**
   * When the select ask to unregister
   */
  @Event() p6FormUnregister!: EventEmitter<P6TranslationControl>;

  @Element() host!: HTMLP6TranslationElement;

  @State() translations: Translation[] = [];

  @State() errorMessage = '';

  private l10n: L10n | undefined;

  async componentWillLoad(): Promise<void> {
    this.l10n = await getL10n(this.host);
    this.host.addEventListener('focusout', this.checkValidity.bind(this));
    this.initializeTranslations();
  }

  /**
   * Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
   * It also triggers the standard error message, such as "this is a required field".
   */
  @Method()
  async validationMessage(): Promise<string> {
    return Promise.resolve(this.errorMessage || '');
  }

  /**
   * Returns whether a form will validate when it is submitted, without having to submit it.
   */
  @Method()
  async checkValidity(): Promise<boolean> {
    const errors = [];
    // default language should have a value
    if (!this.hasDefaultTranslation()) {
      errors.push(
        interpolate(this.l10n?.missingDefaultTranslation, {
          code: getDefaultLanguage(),
          name: getLanguageName(getDefaultLanguage()),
        }),
      );
    }

    const missingTranslations = this.getTranslationsWithoutValue();
    if (missingTranslations.length > 0) {
      errors.push(
        interpolate(this.l10n?.missingTranslation, {
          languages: missingTranslations.map(language => getLanguageName(language)).join(', '),
        }),
      );
    }

    this.errorMessage = errors.join(', ');
    const isValid = this.errorMessage === '';

    if (isValid) {
      this.p6Valid.emit({
        name: this.name,
        value: this.getFieldValue(),
      });
    } else {
      this.p6Invalid.emit({ name: this.name, message: this.errorMessage });
    }

    return Promise.resolve(isValid);
  }

  render(): JSX.Element {
    return (
      <Host>
        <p6-field>
          <p6-label slot="label">
            <slot />
            {!this.readOnly && !this.disabled && (
              <p6-action
                mode={Mode.success}
                size={Size.small}
                onClick={this.onAddLineHandler()}
                disabled={!this.canAddTranslation()}
                class={{
                  'has-tooltip-arrow': true,
                  [`has-tooltip-right`]: true,
                }}
                data-tooltip={this.l10n?.addTooltip}
              >
                <p6-icon name="plus-circle" size={Size.small} />
              </p6-action>
            )}
          </p6-label>
          {this.translations.map(translation => {
            const excludes = this.translations.map(t => t.language).filter(l => l !== translation.language);
            return (
              <p6-translation-line
                language={translation.language}
                translation={translation.value}
                excludes={excludes}
                name={this.name}
                disabled={this.disabled}
                readOnly={this.readOnly}
                onP6Delete={this.onRemoveLineHandler(translation.language)}
                onP6KeyChange={this.onKeyChangeHandler(translation.language)}
                onP6ValueChange={this.onValueChangeHandler(translation.language)}
              />
            );
          })}
        </p6-field>
      </Host>
    );
  }

  componentDidLoad(): void {
    this.p6FormRegister.emit(this);
  }

  disconnectedCallback(): void {
    this.p6FormUnregister.emit(this);
  }

  private initializeTranslations(): void {
    const keys = Object.keys(this.value);
    this.translations = keys.map(language => ({
      language,
      value: this.value[language],
    }));

    if (!keys.includes(getDefaultLanguage())) {
      this.addNewLine(getDefaultLanguage());
    }
  }

  private getTranslationsWithoutValue(): LanguageCode[] {
    return this.translations.filter(translation => translation.language !== '' && translation.value.trim() === '').map(translation => translation.language);
  }

  private getFieldValue(): P6TranslationValue {
    return this.translations.reduce(
      (fieldValue, translation) => ({
        ...fieldValue,
        [translation.language]: translation.value,
      }),
      {} as P6TranslationValue,
    );
  }

  private hasDefaultTranslation(): boolean {
    const defaultTranslation = this.translations.find(translation => translation.language === getDefaultLanguage());
    return defaultTranslation !== undefined && defaultTranslation.value.trim() !== '';
  }

  private canAddTranslation(): boolean {
    return this.translations.length < getLanguageCodes().length;
  }

  private addNewLine(language: LanguageCode): void {
    if (!this.canAddTranslation()) {
      return;
    }
    this.translations = this.translations.concat({
      language,
      value: '',
    });
  }

  private removeLine(language: LanguageCode): void {
    this.translations = this.translations.filter(t => t.language !== language);
  }

  private onRemoveLineHandler(language: LanguageCode): (event: Event) => void {
    return (): void => {
      this.removeLine(language);
    };
  }

  private onAddLineHandler(): (event: Event) => void {
    return (): void => {
      const languages = this.translations.map(translation => translation.language).filter(language => language !== '');
      const missingTranslation = getLanguageCodes().find(language => !languages.includes(language));
      if (missingTranslation !== undefined) {
        this.addNewLine(missingTranslation);
      }
    };
  }

  private onKeyChangeHandler(language: LanguageCode): (event: Event) => void {
    return (event: Event): void => {
      if (!isCustomEvent(event)) {
        return;
      }

      const newLanguage = event.detail.value;
      if (newLanguage === undefined || newLanguage === null) {
        return;
      }

      this.translations = this.translations.map(t => (t.language === language ? { ...t, language: newLanguage } : t));
    };
  }

  private onValueChangeHandler(language: LanguageCode): (event: Event) => void {
    return (event: Event): void => {
      if (!isCustomEvent(event)) {
        return;
      }
      this.translations = this.translations.map(t => (t.language === language ? { ...t, value: event.detail.value } : t));
    };
  }
}
