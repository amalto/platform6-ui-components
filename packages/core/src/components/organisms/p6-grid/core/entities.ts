import { EventEmitter, JSX } from '@stencil/core';
import { Alignment, Direction, SortOrder } from '../../../../shared/types';

export type ColumnId = string;

export type DataItem = Record<string, unknown>;

export interface ColumnDefinition<T extends DataItem> {
  id?: ColumnId;
  label: string;
  field?: string | keyof T;
  editable?: boolean;
  filtreable?: boolean;
  sortable?: boolean;
  hidden?: boolean;
  width?: number;
  align?: Alignment;
  color?: string;
  sortOrder?: SortOrder;
  disableHeaderMenu?: boolean;
  getValue?: (data: T, context: ColumnDefinition<T>) => string;
  setValue?: (value: string, data: T, context: ColumnDefinition<T>) => T;
  cellEditor?: (data: T, context: ColumnDefinition<T>) => JSX.Element & CellEditor<T>;
  cellRenderer?: (data: T, rowId: RowId, context: ColumnDefinition<T>) => string | HTMLElement;
}

export interface Column<T extends DataItem> extends ColumnDefinition<T> {
  id: ColumnId;
  editable: boolean;
  hidden: boolean;
  width: number;
  align: Alignment;
  sortOrder: SortOrder;
  disableHeaderMenu: boolean;
  getValue: (data: T, context: ColumnDefinition<T>) => string;
  setValue: (value: string, data: T, context: ColumnDefinition<T>) => T;
}

export interface CellEditor<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  value: string;

  p6GridCellValueChanged: EventEmitter<CellValueChangedDetail<T>>;
  p6GridEditingCell: EventEmitter<EditingCellDetail<T>>;
}
export type RowId = string;

export interface Row<T extends DataItem> {
  id: RowId;
  data: T;
}

interface ColumnEventDetail<T extends DataItem> {
  column: Column<T>;
}

export interface MoveColumnDetail<T extends DataItem> extends ColumnEventDetail<T> {
  direction: Direction;
}
export type ResizeColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type AlignColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type SortColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type ShowColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export interface CellValueChangedDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  oldValue: string;
}

export enum EditingCellStatus {
  Start,
  Stop,
  Cancel,
}

export interface EditingCellDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  status: EditingCellStatus;
}

export interface ResetDefinitionsDetail {
  reset: boolean;
}

export interface ShowOptionsDetail {
  visible: boolean;
}

export interface FilterRowsDetail {
  value: string;
}
