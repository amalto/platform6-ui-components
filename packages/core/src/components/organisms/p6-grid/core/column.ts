import { Alignment, Direction, SortOrder } from '../../../../shared/types';
import { throwError } from '../../../../utils/error';
import { Column, ColumnDefinition, ColumnId, DataItem } from './entities';

export const MIN_WIDTH = 75;
export const DEFAULT_WIDTH = MIN_WIDTH;
export const INC_WIDTH = 10;

export function defaultGetValue<T extends DataItem>(data: T, context: ColumnDefinition<T>): string {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'getValue' function for ${JSON.stringify(context)}`);
  }

  return `${data[context.field as keyof T]}`;
}

export function defaultSetValue<T extends DataItem>(value: string, data: T, context: ColumnDefinition<T>): T {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'setValue' function for ${JSON.stringify(context)}`);
  }

  return {
    ...data,
    [context.field as keyof T]: value,
  };
}

let columnIds = 0;
export function fromDefinition<T extends DataItem>(column: ColumnDefinition<T>): Column<T> {
  // eslint-disable-next-line no-plusplus
  const id = column.id || column.field || columnIds++;

  return {
    ...column,
    id: `col-${id}`,
    editable: column.editable || false,
    hidden: column.hidden || false,
    width: column.width || 100,
    align: column.align || Alignment.center,
    sortOrder: column.sortOrder || SortOrder.none,
    disableHeaderMenu: column.disableHeaderMenu || false,
    getValue: column.getValue || defaultGetValue,
    setValue: column.setValue || defaultSetValue,
  };
}

export function isHidden<T extends DataItem>(column: Column<T>): boolean {
  return !!column.hidden;
}

export function replaceColumn<T extends DataItem>(updatedColumn: Column<T>, allColumns: Column<T>[]): Column<T>[] {
  return allColumns.map(col => (col.id === updatedColumn.id ? updatedColumn : col));
}

export function move<T extends DataItem>(id: ColumnId, direction: Direction, columns: Column<T>[]): Column<T>[] {
  const from = columns.findIndex(col => col.id === id);
  const to = from + (direction === Direction.left ? -1 : 1);

  if (from < 0 || to < 0 || to >= columns.length) {
    return columns;
  }

  const result = columns.slice();
  const item = result.splice(from, 1);

  return result.slice(0, to).concat(item).concat(result.slice(to, columns.length));
}
