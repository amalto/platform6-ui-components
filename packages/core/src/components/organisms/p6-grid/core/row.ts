import { SortOrder } from '../../../../shared/types';
import { Column, DataItem, Row, RowId } from './entities';

let rowIds = 0;
export function fromData<T extends DataItem>(data: T): Row<T> {
  rowIds += 1;

  return {
    id: `row-${rowIds}`,
    data,
  };
}

export function compareRow<T extends DataItem>(column: Column<T> | undefined): (a: Row<T>, b: Row<T>) => number {
  const sortOrder = column?.sortOrder || SortOrder.none;
  const factor = sortOrder === SortOrder.desc ? -1 : 1;

  return (a: Row<T>, b: Row<T>): number => {
    if (column === undefined || column.sortable === false || sortOrder === SortOrder.none) {
      return 0;
    }

    const valueA = column.getValue(a.data, column);
    const valueB = column.getValue(b.data, column);

    return valueA.localeCompare(valueB) * factor;
  };
}

export function filterBy<T extends DataItem>(searchValue: string, columns: Column<T>[]): (row: Row<T>) => boolean {
  return row => columns.filter(col => col.filtreable !== false).some(col => col.getValue(row.data, col).includes(searchValue));
}

export function rangeSelectRow<T extends DataItem>(currentRowIdSelect: RowId, selectedRows: Set<RowId>, displayedRow: Row<T>[]): Set<RowId> {
  const currentSelectedIndex = displayedRow.findIndex(row => row.id === currentRowIdSelect);

  if (currentSelectedIndex === -1) {
    return selectedRows;
  }

  const previousRowIdSelected = Array.from(selectedRows)
    .reverse()
    .find(rowId => displayedRow.findIndex(row => row.id === rowId) > -1);

  if (previousRowIdSelected === undefined) {
    return new Set(selectedRows.add(currentRowIdSelect));
  }

  const previousSelectedIndex = displayedRow.findIndex(row => row.id === previousRowIdSelected);

  let [from, to] = [previousSelectedIndex, currentSelectedIndex];
  let excludedFromSlice = currentRowIdSelect;
  if (currentSelectedIndex < previousSelectedIndex) {
    [from, to] = [to, from];
    excludedFromSlice = previousRowIdSelected;
  }

  return displayedRow
    .slice(from, to)
    .reduce((selected, cur) => {
      return cur.id === undefined ? selected : selected.add(cur.id);
    }, new Set(selectedRows))
    .add(excludedFromSlice);
}

export function replaceRow<T extends DataItem>(updatedRow: Row<T>, allRows: Row<T>[]): Row<T>[] {
  return allRows.map(row => (row.id === updatedRow.id ? updatedRow : row));
}
