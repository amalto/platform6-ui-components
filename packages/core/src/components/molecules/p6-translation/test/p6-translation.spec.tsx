import { newSpecPage } from '@stencil/core/testing';
import { P6Translation } from '../p6-translation';

describe('p6-translation', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [P6Translation],
      html: `<p6-translation></p6-translation>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
