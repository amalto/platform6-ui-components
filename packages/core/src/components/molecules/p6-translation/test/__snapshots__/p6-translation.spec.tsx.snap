// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-translation renders 1`] = `
<p6-translation>
  <mock:shadow-root>
    <p6-field>
      <p6-label slot="label">
        <slot></slot>
        <p6-action class="has-tooltip-arrow has-tooltip-right" mode="4" size="0">
          <p6-icon name="plus-circle" size="0"></p6-icon>
        </p6-action>
      </p6-label>
      <p6-translation-line language="en" translation=""></p6-translation-line>
    </p6-field>
  </mock:shadow-root>
</p6-translation>
`;
